README.md
Docker-based integration tests
=====
[![PyPI version](https://img.shields.io/pypi/v/pytest-docker?color=green)](https://pypi.org/project/pytest-docker/)
[![Build Status](https://github.com/avast/pytest-docker/actions/workflows/tests.yaml/badge.svg?branch=master)](https://github.com/avast/pytest-docker/actions/workflows/tests.yaml)
[![Python versions](https://img.shields.io/pypi/pyversions/pytest-docker)](https://pypi.org/project/pytest-docker/)
[![Code style](https://img.shields.io/badge/formatted%20with-black-black)](https://github.com/psf/black)


# Description
Simple [pytest](http://doc.pytest.org/) fixtures that help you write integration
tests with Docker and [docker-compose](https://docs.docker.com/compose/).
Specify all necessary containers in a `docker-compose.yml` file and and
`pytest-docker` will spin them up for the duration of your tests.

This package is tested with Python versions `3.6`, `3.7`, `3.8` and
`3.9`, and `pytest` version 4, 5 and 6. Python 2 is not supported.

`pytest-docker` was originally created by André Caron.

# Installation
Install `pytest-docker` with `pip` or add it to your test requirements. It is
recommended to install `docker-compose` python package directly in your
environment to ensure that it is available during tests. This will prevent
potential dependency conflicts that can occur when the system wide
`docker-compose` is used in tests.

The default behavior is not to install `docker-compose` at all. If you
want to, you install `pytest-docker` with the `docker-compose-v1`
extra, you can use the following command:

```
pip install pytest-docker[docker-compose-v1]
```

## Docker Compose v2 compatiblity

This version of `pytest-docker` is compatible with Docker Compose v1
(legacy) and Docker Compose v2 if you have
[`compose-switch`](https://github.com/docker/compose-switch)
installed.

If you want to use the real Docker Compose v2, it has to be installed
system wide ([more
information](https://github.com/docker/compose#linux)).

# Usage
Here is an example of a test that depends on a HTTP service.

With a `docker-compose.yml` file like this (using the
[httpbin](https://httpbin.org/) service):

```yaml
version: '2'
services:
  httpbin:
    image: "kennethreitz/httpbin"
    ports:
      - "8000:80"
```

You can write a test like this:

```python
import pytest
import requests

from requests.exceptions import ConnectionError


def is_responsive(url):
    try:
        response = requests.get(url)
        if response.status_code == 200:
            return True
    except ConnectionError:
        return False


@pytest.fixture(scope="session")
def http_service(docker_ip, docker_services):
    """Ensure that HTTP service is up and responsive."""

    # `port_for` takes a container port and returns the corresponding host port
    port = docker_services.port_for("httpbin", 80)
    url = "http://{}:{}".format(docker_ip, port)
    docker_services.wait_until_responsive(
        timeout=30.0, pause=0.1, check=lambda: is_responsive(url)
    )
    return url


def test_status_code(http_service):
    status = 418
    response = requests.get(http_service + "/status/{}".format(status))

    assert response.status_code == status
```

By default this plugin will try to open `docker-compose.yml` in your
`tests` directory. If you need to use a custom location, override the
`docker_compose_file` fixture inside your `conftest.py` file:

```python
import os
import pytest


@pytest.fixture(scope="session")
def docker_compose_file(pytestconfig):
    return os.path.join(str(pytestconfig.rootdir), "mycustomdir", "docker-compose.yml")
```

## Available fixtures
All fixtures have `session` scope.

### `docker_ip`

Determine the IP address for TCP connections to Docker containers.

### `docker_compose_file`

Get an absolute path to the  `docker-compose.yml` file. Override this fixture in
your tests if you need a custom location.

### `docker_compose_project_name`

Generate a project name using the current process PID. Override this fixture in
your tests if you need a particular project name.

### `docker_services`

Start all services from the docker compose file (`docker-compose up`).
After test are finished, shutdown all services (`docker-compose down`).

### `docker_cleanup`

Get the docker compose command to execute for test clean-up actions. Override
this fixture in your tests if you need custom clean-up actions.

### `docker_compose_command`

Docker Compose command to use to execute Dockers. Default is to use
Docker Compose v1 (command is `docker-compose`). If you want to use
Docker Compose v2, change this fixture to return `docker compose`.

# Development
Use of a virtual environment is recommended. See the
[venv](https://docs.python.org/3/library/venv.html) package for more
information.

First, install `pytest-docker` and its test dependencies:

	pip install -e ".[tests]"

Run tests with

	pytest -c setup.cfg

to make sure that the correct configuration is used. This is also how tests are
run in CI.

Use [black](https://pypi.org/project/black/) with default settings for
formatting. You can also use `pylint` with `setup.cfg` as the configuration
file.


# Contributing
This pytest plug-in and its source code are made available to you under a MIT
license. It is safe to use in commercial and closed-source applications. Read
the license for details!

Found a bug? Think a new feature would make this plug-in more practical? We
welcome issues and pull requests!

When creating a pull request, be sure to follow this projects conventions (see
above).


CHANGELOG.md
# Changelog

## Unreleased
Changes:
- Default behavior is not to install `docker-compose` at all when
  installing `pytest-docker` with PIP. If you want to, you install
  `pytest-docker` with the `docker-compose-v1` extra (`pip install
  pytest-docker[docker-compose-v1]`).
- Add support for Docker Compose v2 via a new pytest fixture,
  `docker_compose_command` that should return `docker compose`.

## Version 0.10.3
Changes:
- Ensure that Docker cleanup is executed even with after tests have failed

## Version 0.10.2
Changes:
- Allow higher version of `attrs` (21.0)

## Version 0.10.1
Changes:
- Allow higher version of `attrs`

## Version 0.10.0
Changes:
- Drop Python3.5 support

## Version 0.9.0
Changes:
- Add the `docker_cleanup` fixture to allow custom cleanup actions for
docker-compose

## Version 0.8.0
Changes:
- Add explicit dependencies on `docker-compose` and `pytest`
- Stop using deprecated `pytest-runner` to run package tests

## Version 0.7.2
Changes:
- Update package README

## Version 0.7.1
- Minor packaging related fixes
	- Fix description content type in `setup.cfg`
	- Add `skip_cleanup` option to travis-ci settings

## Version 0.7.0
- Removed the fallback method that allowed running without Docker
- Package cleanup

## Version 0.6.0
- Added ability to return list of files from `docker_compose_file` fixture

## Version 0.3.0
- Added `--build` option to `docker-compose up` command to automatically
  rebuild local containers


tests/test_fixtures.py
import os.path


HERE = os.path.dirname(os.path.abspath(__file__))


def test_docker_compose_file(docker_compose_file):
    assert docker_compose_file == os.path.join(HERE, "docker-compose.yml")


def test_docker_compose_project(docker_compose_project_name):
    assert docker_compose_project_name == "pytest{}".format(os.getpid())


def test_docker_cleanup(docker_cleanup):
    assert docker_cleanup == "down -v"


def test_docker_compose_comand(docker_compose_command):
    assert docker_compose_command == "docker-compose"


tests/test_docker_services.py
import subprocess
from unittest import mock

import pytest
from pytest_docker.plugin import (
    DockerComposeExecutor,
    Services,
    get_docker_services,
    get_cleanup_command,
)


def test_docker_services():
    """Automatic teardown of all services."""

    with mock.patch("subprocess.check_output") as check_output:
        check_output.side_effect = [b"", b"0.0.0.0:32770", b""]
        check_output.returncode = 0

        assert check_output.call_count == 0

        # The fixture is a context-manager.
        with get_docker_services(
            "docker-compose",
            "docker-compose.yml",
            docker_compose_project_name="pytest123",
            docker_cleanup=get_cleanup_command(),
        ) as services:
            assert isinstance(services, Services)

            assert check_output.call_count == 1

            # Can request port for services.
            port = services.port_for("abc", 123)
            assert port == 32770

            assert check_output.call_count == 2

            # 2nd request for same service should hit the cache.
            port = services.port_for("abc", 123)
            assert port == 32770

            assert check_output.call_count == 2

        assert check_output.call_count == 3

    # Both should have been called.
    assert check_output.call_args_list == [
        mock.call(
            'docker-compose -f "docker-compose.yml" -p "pytest123" up --build -d',
            stderr=subprocess.STDOUT,
            shell=True,
        ),
        mock.call(
            'docker-compose -f "docker-compose.yml" -p "pytest123" port abc 123',
            stderr=subprocess.STDOUT,
            shell=True,
        ),
        mock.call(
            'docker-compose -f "docker-compose.yml" -p "pytest123" down -v',
            stderr=subprocess.STDOUT,
            shell=True,
        ),
    ]


def test_docker_services_unused_port():
    """Complain loudly when the requested port is not used by the service."""

    with mock.patch("subprocess.check_output") as check_output:
        check_output.side_effect = [b"", b"", b""]
        check_output.returncode = 0

        assert check_output.call_count == 0

        # The fixture is a context-manager.
        with get_docker_services(
            "docker-compose",
            "docker-compose.yml",
            docker_compose_project_name="pytest123",
            docker_cleanup=get_cleanup_command(),
        ) as services:
            assert isinstance(services, Services)

            assert check_output.call_count == 1

            # Can request port for services.
            with pytest.raises(ValueError) as exc:
                print(services.port_for("abc", 123))
            assert str(exc.value) == (
                'Could not detect port for "%s:%d".' % ("abc", 123)
            )

            assert check_output.call_count == 2

        assert check_output.call_count == 3

    # Both should have been called.
    assert check_output.call_args_list == [
        mock.call(
            'docker-compose -f "docker-compose.yml" -p "pytest123" ' "up --build -d",
            shell=True,
            stderr=subprocess.STDOUT,
        ),
        mock.call(
            'docker-compose -f "docker-compose.yml" -p "pytest123" ' "port abc 123",
            shell=True,
            stderr=subprocess.STDOUT,
        ),
        mock.call(
            'docker-compose -f "docker-compose.yml" -p "pytest123" down -v',
            shell=True,
            stderr=subprocess.STDOUT,
        ),
    ]


def test_docker_services_failure():
    """Propagate failure to start service."""

    with mock.patch("subprocess.check_output") as check_output:
        check_output.side_effect = [
            subprocess.CalledProcessError(1, "the command", b"the output")
        ]
        check_output.returncode = 1

        # The fixture is a context-manager.
        with pytest.raises(Exception) as exc:
            with get_docker_services(
                "docker-compose",
                "docker-compose.yml",
                docker_compose_project_name="pytest123",
                docker_cleanup=get_cleanup_command(),
            ):
                pass

        # Failure propagates with improved diagnoatics.
        assert str(exc.value) == (
            'Command {} returned {}: """{}""".'.format("the command", 1, "the output")
        )

        assert check_output.call_count == 1

    # Tear down code should not be called.
    assert check_output.call_args_list == [
        mock.call(
            'docker-compose -f "docker-compose.yml" -p "pytest123" ' "up --build -d",
            shell=True,
            stderr=subprocess.STDOUT,
        )
    ]


def test_wait_until_responsive_timeout():
    clock = mock.MagicMock()
    clock.side_effect = [0.0, 1.0, 2.0, 3.0]

    with mock.patch("time.sleep") as sleep:
        docker_compose = DockerComposeExecutor(
            compose_command="docker-compose",
            compose_files="docker-compose.yml",
            compose_project_name="pytest123",
        )
        services = Services(docker_compose)
        with pytest.raises(Exception) as exc:
            print(
                services.wait_until_responsive(
                    check=lambda: False, timeout=3.0, pause=1.0, clock=clock
                )
            )
        assert sleep.call_args_list == [mock.call(1.0), mock.call(1.0), mock.call(1.0)]
    assert str(exc.value) == ("Timeout reached while waiting on service!")


src/pytest_docker/plugin.py
import contextlib
import os
import re
import subprocess
import time
import timeit

import attr

import pytest


def execute(command, success_codes=(0,)):
    """Run a shell command."""
    try:
        output = subprocess.check_output(command, stderr=subprocess.STDOUT, shell=True)
        status = 0
    except subprocess.CalledProcessError as error:
        output = error.output or b""
        status = error.returncode
        command = error.cmd

    if status not in success_codes:
        raise Exception(
            'Command {} returned {}: """{}""".'.format(
                command, status, output.decode("utf-8")
            )
        )
    return output


def get_docker_ip():
    # When talking to the Docker daemon via a UNIX socket, route all TCP
    # traffic to docker containers via the TCP loopback interface.
    docker_host = os.environ.get("DOCKER_HOST", "").strip()
    if not docker_host:
        return "127.0.0.1"

    match = re.match(r"^tcp://(.+?):\d+$", docker_host)
    if not match:
        raise ValueError('Invalid value for DOCKER_HOST: "%s".' % (docker_host,))
    return match.group(1)


@pytest.fixture(scope="session")
def docker_ip():
    """Determine the IP address for TCP connections to Docker containers."""

    return get_docker_ip()


@attr.s(frozen=True)
class Services:

    _docker_compose = attr.ib()
    _services = attr.ib(init=False, default=attr.Factory(dict))

    def port_for(self, service, container_port):
        """Return the "host" port for `service` and `container_port`.

        E.g. If the service is defined like this:

            version: '2'
            services:
              httpbin:
                build: .
                ports:
                  - "8000:80"

        this method will return 8000 for container_port=80.
        """

        # Lookup in the cache.
        cache = self._services.get(service, {}).get(container_port, None)
        if cache is not None:
            return cache

        output = self._docker_compose.execute("port %s %d" % (service, container_port))
        endpoint = output.strip().decode("utf-8")
        if not endpoint:
            raise ValueError(
                'Could not detect port for "%s:%d".' % (service, container_port)
            )

        # This handles messy output that might contain warnings or other text
        if len(endpoint.split("\n")) > 1:
            endpoint = endpoint.split("\n")[-1]

        # Usually, the IP address here is 0.0.0.0, so we don't use it.
        match = int(endpoint.split(":", 1)[1])

        # Store it in cache in case we request it multiple times.
        self._services.setdefault(service, {})[container_port] = match

        return match

    def wait_until_responsive(self, check, timeout, pause, clock=timeit.default_timer):
        """Wait until a service is responsive."""

        ref = clock()
        now = ref
        while (now - ref) < timeout:
            if check():
                return
            time.sleep(pause)
            now = clock()

        raise Exception("Timeout reached while waiting on service!")


def str_to_list(arg):
    if isinstance(arg, (list, tuple)):
        return arg
    return [arg]


@attr.s(frozen=True)
class DockerComposeExecutor:

    _compose_command = attr.ib()
    _compose_files = attr.ib(converter=str_to_list)
    _compose_project_name = attr.ib()

    def execute(self, subcommand):
        command = self._compose_command
        for compose_file in self._compose_files:
            command += ' -f "{}"'.format(compose_file)
        command += ' -p "{}" {}'.format(self._compose_project_name, subcommand)
        return execute(command)


@pytest.fixture(scope="session")
def docker_compose_command():
    """Docker Compose command to use, it could be either `docker-compose`
    for Docker Compose v1 or `docker compose` for Docker Compose
    v2."""

    return "docker-compose"


@pytest.fixture(scope="session")
def docker_compose_file(pytestconfig):
    """Get an absolute path to the  `docker-compose.yml` file. Override this
    fixture in your tests if you need a custom location."""

    return os.path.join(str(pytestconfig.rootdir), "tests", "docker-compose.yml")


@pytest.fixture(scope="session")
def docker_compose_project_name():
    """Generate a project name using the current process PID. Override this
    fixture in your tests if you need a particular project name."""

    return "pytest{}".format(os.getpid())


def get_cleanup_command():

    return "down -v"


@pytest.fixture(scope="session")
def docker_cleanup():
    """Get the docker_compose command to be executed for test clean-up actions.
     Override this fixture in your tests if you need to change clean-up actions."""

    return get_cleanup_command()


@contextlib.contextmanager
def get_docker_services(
    docker_compose_command,
    docker_compose_file,
    docker_compose_project_name,
    docker_cleanup,
):
    docker_compose = DockerComposeExecutor(
        docker_compose_command, docker_compose_file, docker_compose_project_name
    )

    # Spawn containers.
    docker_compose.execute("up --build -d")

    try:
        # Let test(s) run.
        yield Services(docker_compose)
    finally:
        # Clean up.
        docker_compose.execute(docker_cleanup)


@pytest.fixture(scope="session")
def docker_services(
    docker_compose_command,
    docker_compose_file,
    docker_compose_project_name,
    docker_cleanup,
):
    """Start all services from a docker compose file (`docker-compose up`).
    After test are finished, shutdown all services (`docker-compose down`)."""

    with get_docker_services(
        docker_compose_command,
        docker_compose_file,
        docker_compose_project_name,
        docker_cleanup,
    ) as docker_service:
        yield docker_service
